File: brady/balanced_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // O(n) time, O(n) space
    public boolean isBalanced(TreeNode root) {
        int[] res = heightAndWhetherBalanced(root);

        return res[1] == 0;
    }

    // Second value in the array will be a 1 if the recursive calls found a height difference > 1, 0 otherwise
    public int[] heightAndWhetherBalanced(TreeNode root) {
        if (root == null) return new int[]{0, 0};

        int[] leftRes = heightAndWhetherBalanced(root.left);
        int[] rightRes = heightAndWhetherBalanced(root.right);
        int heightDifference = Math.abs(leftRes[0]-rightRes[0]);
        int foundImbalance = (leftRes[1] == 1 || rightRes[1] == 1 || heightDifference > 1) ? 1 : 0;

        return new int[]{Math.max(leftRes[0], rightRes[0])+1, foundImbalance};
    }
}

File: gabe/balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def utils(root):

            if root == None:
                return 0

            left = utils(root.left)

            if left == -1:
                return -1 

            right = utils(root.right)

            if right == -1 or abs(right-left) > 1:
                return -1
            else:
                return max(right, left) + 1

        return False if utils(root) == -1 else True